Add support for new Indian regulations on recurring payments

reykjalin opened the pull request:
# Changes proposed in this Pull Request

The changes in this PR are intended to add support for new 3DS regulations for cards issued in India.

- Migrate all card payments to the Payment Method API from the Source API
- Add necessary information to API requests for subscription payment that allow Stripe to follow Indian 3DS regulations
- Add handling for specific payment intent statuses and errors that can occur during the new Indian 3DS flow


# Known Issues

<details><summary>There are several known issues with the current implementation, expand for details</summary>

- If you save a card before applying the changes included in this PR it will not work with the new Indian 3DS regulations.
    - Manual fix: delete the saved payment method from your account and add it again.
    - Repro steps:
        - Set your store currency to INR
        - `git switch develop`
        - `npm run start:webpack` (you can exit this as soon as it's finished bundling the JS)
        - Buy a regular product with the Indian 3DS mandate test card: `4000003560000008` and make sure to save the card
        - `git switch fix/indian-3ds-recurring-regulation-change`
        - `npm run start:webpack` (you can exit this as soon as it's finished bundling the JS)
        - Buy a subscription using the card you saved earlier
        - Go to **Tools** > **Scheduled Actions** > **Pending**
        - Locate the renewal task for the subscription you just bought
        - Click **Run** to run the renewal payment
        - Notice that the payment is immediately successful in your Stripe dashboard when it should take ~5 minutes to transition from `requires_action` to `succeeded`
    - Suggested fix: Not sure. Transition the card somehow from being a `src_xxx` to `pm_xxx`? Not sure if that's possible though.
- Changing sub payment method from one that requires a mandate (`4000003560000008`) to one that does not (e.g. `4000000000003220`) is not possible.
    - Repro steps:
        - Purchase a subscription with Indian 3DS mandate test card `4000003560000008`
        - Go to **Tools** > **Scheduled Actions** > **Pending**
        - Locate the renewal task for the subscription you just bought
        - Click **Run** to run the renewal payment
        - Wait ~5 minutes to make sure the renewal payment is successful (order will transition from `pending` to `processing`)
        - Go to **My Account** > **Subscriptions**, find the subscription you just purchased, and change the payment method to a non-India 3DS mandate test card, e.g. `4000000000003220`.
        - Go to **Tools** > **Scheduled Actions** > **Pending**
        - Locate the renewal task for the subscription you just bought
        - Click **Run** to run the renewal payment
        - Go to **WooCommerce** > **Orders**
        - The latest renewal order should have failed
        - The reason is because the mandate is invalid
    - Suggested fix: clear the `_stripe_mandate_id` meta when no `mandate` ID is included in the API response so we don't use a mandate when it's not required.
        - Might not be a good idea if renewal payments don't include the mandate in the response? 🤔 
- Saved cards don't work with the UPE element
    - Repro steps:
        - Purchase any product with `4000003560000008` and make sure the card is saved.
        - Use the card you saved above to buy a subscription.
        - Go to **Tools** > **Scheduled Actions** > **Pending**
        - Locate the renewal task for the subscription you just bought
        - Click **Run** to run the renewal payment
        - Notice that the expected asynchronous order update did not happen; it was successful immediately.
        - If you look at Stripe's payment request log you'll see that the API didn't actually create a `mandate` for the payment.
    - Suggested fix: Don't know; maybe need @FangedParakeet to weigh in on that? I really don't know why this fails; I have absolutely no clue.
    - **~~There is the possibility~~[It is very likely](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981#issuecomment-939246499) that there is some problem with Stripe's testing API. It has happened before.**
- Including https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1634 since testing instructions require running through all critical flows.

</details>


# Testing instructions

> **Note:** You must set the currency of your store to Indian Rupees (INR) before testing the changes in this PR.

> **Note:** This PR relies on webhooks. You must configure your test site such that it can receive webhooks for these tests to work as intended.

## Test all critical flows

> Please test these flows with your store currency set to INR and then go through the flows again with a different currency.

Make sure all [critical flows](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows) work as intended with both the Classic and UPE checkouts. Make sure to test both Shortcode and Block-based flows.

Try to go through the critical flows with a normal card, 3DS card, and the "success" test Indian regulatory card `4000003560000008`.

## Successful subscription purchase and renewal

1. Buy a subscription with `4000003560000008`
2. Go to **Tools** > **Scheduled Actions** > **Pending**
3. Locate the renewal task for the subscription you just bought
4. Click **Run** to run the renewal payment
5. Go to **WooCommerce** > **Orders**
6. Open the new renewal order
7. Make sure there's an order notes saying something along the lines of "The customer must authorize this payment off-session. An attempt will be made to charge the customer’s card on 2021-10-10 at 04:09."
8. Wait ~5 minutes
9. Reload the order details page
10. Make sure the payment has been successfully processed


## Successful subscription purchase and renewal with multiple subscriptions on a different schedule

1. Create 2 subscription products: one with a daily renewal schedule, the other with a weekly renewal schedule.
2. Buy both subscriptions in a single order with `4000003560000008`.
3. Follow steps 2-10 in the **Successful subscription purchase and renewal** for the daily renewal subscription.
4. Follow steps 2-10 in the **Successful subscription purchase and renewal** for the weekly renewal subscription.


## Successful subscription purchase and renewal with multiple subscriptions on the same schedule

1. Create 2 subscription products, both with the same renewal schedule.
2. Buy both subscriptions in a single order with `4000003560000008`.
3. Follow steps 2-10 in the **Successful subscription purchase and renewal** for the daily renewal subscription.
4. Follow steps 2-10 in the **Successful subscription purchase and renewal** for the weekly renewal subscription.


## Successful subscription purchase with non-subscription product in the cart

1. Buy a subscription and any amount of non-subscription items with `4000003560000008`.
2. Go to **Tools** > **Scheduled Actions** > **Pending**
3. Locate the renewal task for the subscription you just bought
4. Click **Run** to run the renewal payment
5. Go to **WooCommerce** > **Orders**
6. Open the new renewal order
7. Make sure there's an order notes saying something along the lines of "The customer must authorize this payment off-session. An attempt will be made to charge the customer’s card on 2021-10-10 at 04:09."
8. Wait ~5 minutes
9. Reload the order details page
10. Make sure the payment has been successfully processed


## Successful subscription purchase, but failed renewal

> **Note:** The order note explaining the failure is a generic "payment failed" note. Adding a better note should be handled in a different PR.

1. Buy a subscription with `4000003560000099`
2. Go to **Tools** > **Scheduled Actions** > **Pending**
3. Locate the renewal task for the subscription you just bought
4. Click **Run** to run the renewal payment
5. Go to **WooCommerce** > **Orders**
6. Open the new renewal order
7. Make sure there's an order note explaining the order failed
8. Make sure the order status is set to **Failed**.

Now repeat those steps, but buy a subscription with `4000003560000040` in step 1.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
Add required information to intent creation

Stripe will use this information to create a "mandate" for Indian cards; the mandate

will be used to notify the customer and have them authorize the subscription

off-session.

There are currently *a lot* of unknowns:

 - What happens when an order has multiple subscriptions?

 - What happens when an order has multiple subs, all with different payment periods?

The current code hasn't been tested at all. The payment intent updates also haven't

been implemented, so this wouldn't work anyway.

Next up:

 - Get test card numbers from Stripe

 - Ask Stripe about unknowns ^

 - Handle payment intent updates via webhooks

Commit:
Fix fatal error in subs trait

The number of arguments passed to the new filter was missing

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -62,6 +65,50 @@ public function maybe_init_subscriptions() {
+		$sub = $subscriptions[0];
+
+		// TODO: Maybe the order ID isn't unique enough? What happens when there are multiple subscriptions
+		// with different payment intervals in the order?
```
I think the order ID is unique enough. In the case of multiple subs in a single order we'll likely want the mandate associated with the original order instead of one of the subscriptions.

reykjalin replied:
The order ID is enough since the mandate should be associated with the subscription in general — not specific orders.

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -62,6 +65,50 @@ public function maybe_init_subscriptions() {
+		// with different payment intervals in the order?
+		$request['payment_method_options']['card']['mandate_options']['reference']      = $order->get_id();
+		$request['payment_method_options']['card']['mandate_options']['amount']         = $order->get_total();
+		$request['payment_method_options']['card']['mandate_options']['amount_type']    = 'fixed'; // TODO: Are there any cases where amount is variable?
```
I think the amount should be `maximum` whenever a single order contains multiple subscriptions so we can use the mandate to charge each individual subscription separately.

Maybe `maximum` is just a better approach overall?

ismaeldcom replied:
I still need to look at multiple subscriptions. But if we reuse the mandate for an order with multiple subscription and then renew only one of them, shouldn't the `amount_type` be `maximum`?

reykjalin replied:
Yep, that's how I understand the instructions from Stripe 🙂 

Haven't done any sort of testing to verify that though.

reykjalin replied:
I've verified that setting the amount type to `maximum` and interval to `sporadic` works as intended here.

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -62,6 +65,50 @@ public function maybe_init_subscriptions() {
+		$request['payment_method_options']['card']['mandate_options']['reference']      = $order->get_id();
+		$request['payment_method_options']['card']['mandate_options']['amount']         = $order->get_total();
+		$request['payment_method_options']['card']['mandate_options']['amount_type']    = 'fixed'; // TODO: Are there any cases where amount is variable?
+		$request['payment_method_options']['card']['mandate_options']['start_date']     = $sub->get_date( 'start', 'gmt' ); // TODO: Better to use site timezone?
```
I'm not sure here. Does the timezone ever matter?

ismaeldcom replied:
🤔 Not sure either, it's `gmt` by default. I think you can simplify the latest code using `get_time` instead of `get_date `.

reykjalin replied:
Oooh, I didn't even know that function existed! Thanks for the suggestion, change in a8525e33e7a5ea654fe3b1e77722c619a0512b87 🎉

Commit:
Fix fatal PHP errors and API errors

Commit:
Add mandate ID to order metadata when available

Commit:
Use pre-existing mandate when appropriate

Commit:
Retrieve mandate for renewal payments

Not sure if the previous approach is still required, since this is

technically only required for renewal payments.

Commit:
Inform merchant when async payment auth required

This is a completely blind attempt at trying to make this work; I have
no clue how this will end up working.

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -319,6 +393,23 @@ public function process_subscription_payment( $amount, $renewal_order, $retry =
 				throw new WC_Stripe_Exception( print_r( $response, true ), $localized_message );
 			}
 
+			if ( ! empty( $response->next_action ) && 'card_await_notification' === $response->next_action->type ) {
+				$charge_attempt_at = $response->next_action->card_await_notification->charge_attempt_at;
+				$date              = new DateTime();
+				$date->setTimestamp( $charge_attempt_at );
+				$date->setTimezone( wp_timezone() );
+				$charge_attempt_date_string = $date->format( 'Y-m-d' );
+				$charge_attempt_time_string = $date->format( 'H:i' );
+
+				$message = sprintf(
+					/* translators: 1) a date in the format yyyy-mm-dd, e.g. 2021-09-21; 2) time in the 24-hour format HH:mm, e.g. 23:04 */
+					__( 'The customer must authorize this payment off-session. An attempt will be made to charge the customer\'s card on %1$s at %1$s.', 'woocommerce-gateway-stripe' ),
+					$charge_attempt_date_string,
+					$charge_attempt_time_string
+				);
+				$renewal_order->add_order_note( $message );
+			}
+
```
This is a big guess on the implementation of this; I have no idea if this will actually work since there is currently no test card available to test this.

We're entirely reliant on Stripe to add test card numbers so we can figure out how best to handle this.

reykjalin replied:
This was the right approach, but has been tweaked to be more robust.

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -62,6 +65,77 @@ public function maybe_init_subscriptions() {
+			return $request;
+		}
+
+		// TODO: what happens if there's more than one renewal order?
+		$renewals = wcs_get_subscriptions_for_renewal_order( $order->get_id() );
+		if ( 1 === count( $renewals ) ) {
+			$renewal_order   = reset( $renewals );
+			$parent_order_id = $renewal_order->get_parent_id();
+			$parent_order    = wc_get_order( $parent_order_id );
+
+			$mandate = $parent_order->get_meta( '_stripe_mandate_id', true );
+			if ( isset( $mandate ) && '' !== $mandate ) {
+				$request['mandate'] = $mandate;
+				return $request;
+			}
+		}
```
I think in this case it's fine that there's more than one order; they'd all presumably have the same mandate, so this is probably fine as is 🤔

reykjalin replied:
This has been updated to be more robust, but the answer to the previous question is: yes, all renewal orders will use the same mandate.

Commit:
Add support for multiple orders

I'm not sure if this is the right approach since the orders actually
don't fail in test mode when the mandate options are configured
incorrectly. As such, there is no way to verify whether this is the
correct approach or not.

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -62,6 +65,80 @@ public function maybe_init_subscriptions() {
+		// Get the first subscription associated with this order.
+		$sub = reset( $subscriptions );
+
+		if ( 1 === count( $subscriptions ) ) {
+			$request['payment_method_options']['card']['mandate_options']['amount_type']    = 'fixed';
+			$request['payment_method_options']['card']['mandate_options']['interval']       = $sub->get_billing_period();
+			$request['payment_method_options']['card']['mandate_options']['interval_count'] = $sub->get_billing_interval();
+		} else {
+			// If there are multiple subscriptions the amount_type becomes 'maximum' so we can charge anything
+			// less than the order total, and the interval is sporadic so we don't have to follow a set interval.
+			$request['payment_method_options']['card']['mandate_options']['amount_type'] = 'maximum';
+			$request['payment_method_options']['card']['mandate_options']['interval']    = 'sporadic';
+		}
```
I'm not sure if this is the right approach since the orders actually don't fail in test mode when the mandate options are configured incorrectly. As such, there is no way to verify whether this is the correct approach or not.

We'll have to wait for further direction from Stripe so we can test this.

reykjalin replied:
This has been verified to be the correct approach.

Commit:
Add initial support for the payment methods API

This only applies to card payments and is necessary to support the new
Indian 3DS regulations since those are not supported by the Sources API.

Commit:
Add informative order note when mandate is invalid

This is instead of just a generic "can't process this payment" order
note.

Commit:
Save card when paying for subscription

Commit:
Handle new checkout via payment methods properly

Instead of using the UPE gateway like before we now support Payment
Methods with the base gateway.

Commit:
Fix note when waiting for pending authorization

Makes sure only the note informing the merchant is shown, instead of a
note asking them to go process the payment in their dashboard.

Commit:
Add support for adding a payment method

In the context of non-checkout, i.e. My Account > Payment Methods

Commit:
Ensure mandate param is only incl when appropriate

There were edge cases where the mandate parameter was passed with an API
request where "confirm" was not set to "true".

Commit:
Switch to payment method api for blocks checkout

Commit:
Fix save card handling

There were some assumptions in the code that weren't being met when
using payment methods. This has now been fixed by correcting several if
statements and their assumptions.

Commit:
Fix type in comment

reykjalin commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -552,7 +552,13 @@ public function process_payment( $order_id, $retry = true, $force_save_source =
+			) {
 				$upe_gateway = new WC_Stripe_UPE_Payment_Gateway();
 				return $upe_gateway->process_payment_with_saved_payment_method( $order_id );
 			}
```
@diegocurbelo can we remove this check now that the classic gateway can process payments made with Payment Methods, or do we still defer to the UPE gateway when it's enabled?

Asking you specifically since the git blame mentions you on this if statement 🙂

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -62,6 +65,80 @@ public function maybe_init_subscriptions() {
+		if ( 'true' === $request['confirm'] && isset( $mandate ) && '' !== $mandate ) {
+			$request['mandate'] = $mandate;
+			return $request;
+		}
```
I think this check is required; it's a good failsafe at the very least. Would love some thoughts on it though.

Some context:

Initial mandate ID is saved to the parent order, and this function will receive renewal orders most of the time. This check is here just in case we ever receive the parent order for whatever reason. I'm not sure if this can actually happen, but probably better safe than sorry?

ricardo replied:
Perhaps we can simplify `isset( $mandate ) && '' !== $mandate` to `! empty( $mandate )` in case it can't be `0`.

reykjalin replied:
Yep, the mandate should always be some form of `'mandate_1234'` 🙂 
Excellent suggestion! Applied this in 025fe3fe1367774a8fec88f0d38685d1ea1ed14f

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -62,6 +65,80 @@ public function maybe_init_subscriptions() {
+		$request['payment_method_options']['card']['mandate_options']['reference']  = $order->get_id();
+		$request['payment_method_options']['card']['mandate_options']['start_date'] = $start_date->getTimestamp();
+
+		// TODO: add optional mandate options? description, end_date.
```
This is far from necessary, so we can think about this in a separate issue.

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -310,6 +387,11 @@ public function process_subscription_payment( $amount, $renewal_order, $retry =
 
 				if ( 'card_error' === $response->error->type ) {
 					$localized_message = isset( $localized_messages[ $response->error->code ] ) ? $localized_messages[ $response->error->code ] : $response->error->message;
+				} elseif ( 'payment_intent_mandate_invalid' === $response->error->code ) {
+					$localized_message = __(
+						'The mandate used for this renewal payment is invalid. You may need to bring the customer back to your store and ask them to resubmit their payment information.',
+						'woocommerce-gateway-stripe'
+					);
```
@LevinMedia what do you think about the phrasing here?

**Context**

A mandate is required to process renewal payments for subscriptions with cards issued in India. We create a mandate when a subscription is created. A mandate can be invalid for any number of reasons, e.g. if we try to charge the wrong amount which could happen if a merchant changes the price of their subscription.

This can happen for multiple reasons so we can't really offer a proper way to address the issue. However, the most likely reason is that something in the subscription changed which means the customer will have to come back on-session to re-authorize the payment, which is the reason I wanted to include the "You may need to bring the customer back to your store and ask them to resubmit their payment information" part of the order note.

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -336,6 +418,25 @@ public function process_subscription_payment( $amount, $renewal_order, $retry =
+
+				$message = sprintf(
+					/* translators: 1) a date in the format yyyy-mm-dd, e.g. 2021-09-21; 2) time in the 24-hour format HH:mm, e.g. 23:04 */
+					__( 'The customer must authorize this payment off-session. An attempt will be made to charge the customer\'s card on %1$s at %2$s.', 'woocommerce-gateway-stripe' ),
```
@LevinMedia What do you think about the phrasing here?

**Context**

With the new regulations for 3DS payments with cards issued in India customers must authorize renewal payments off-session. The card issuer will send the customer a prompt (email/sms) to authorize the payment. The authorization happens through the card issuer's system not the online store. Stripe doesn't know when (or if) the customer will authorize the payment and will instead attempt to charge the card after a certain waiting period.

The order will have a status of `pending` while waiting for authorization.

Some things to think about:

- Is the note clear enough to tell the merchant what's going on?
- Is date and time too much information? Is just the date sufficient?

ismaeldcom replied:
I'm interested in this too. For WC Pay I used `The customer must authorize this payment via the pre-debit notification sent to them by their card issuing bank, before %1$s at %2$s, when the charge will be attempted.` with the date and time formatted using the store settings. But would like to know your thoughts @LevinMedia. Maybe we can agree on the same text for both plugins.

LevinMedia replied:
@reykjalin I like the copy that @ismaeldcom is working with, although  "Pre-Debit" seems a bit Jargony... 

Could we update the copy in both places to read: 

>The customer must authorize this payment via a notification sent to them by the bank which issued their card. The authorization must be completed before %1$s at %2$s, when the charge will be attempted.

ricardo replied:
> The customer must authorize this payment off-session. An attempt will be made to charge the customer’s card on 2021-10-12 at 01:19.

FWIW I was a bit confused with the part where it says the "An attempt will be made to charge" in the future tense, while mentioning a fixed timestamp of when the renewal process was dispatched. - I wasn't sure if the charge would be attempted immediately or if it was a timezone offset thing.

reykjalin replied:
> I was a bit confused with the part where it says the "An attempt will be made to charge" in the future tense, while mentioning a fixed timestamp of when the renewal process was dispatched

The timestamp is supposed to be in the future; probably a time zone issue since `$renewal->get_time( 'start' )` defaults to `gmt` as the timezone. We should probably just pass in the site's timezone like so: `$renewal->get_time( 'start', wp_timezone() )` 🤔 

There's another discussion on this further up in this PR: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981#discussion_r718054458

reykjalin replied:
I've updated the order note in 195d496bead674c2a9c42b197323fb5840e3f01f

Commit:
Update TODOs and comments

Commit:
Use get_time instead of get_date for sub start

Commit:
Add changelog entry

ismaeldcom reviewed (dismissed):
Thanks @reykjalin for your work here! I followed your testing instructions, and they work as expected. 🎉 

Checked the code, and without taking into account the differences with WC Pay, all the common parts are pretty similar, and nothing strange stands out.

**I'm approving this, but would be better to have another approval from a member of your team, more familiarized with the plugin** 🙂

ricardo commented on assets/js/stripe.js:
```diff
@@ -505,6 +505,13 @@ jQuery( function( $ ) {
 				extra_details.type     = 'sepa_debit';
 
 				return stripe.createSource( iban, extra_details ).then( wc_stripe_form.sourceResponse );
+			} else if ( wc_stripe_form.isStripeCardChosen() ) {
```
Since this works only for cards and SEPA, we probably don't need the `return stripe.createSource` at the end. (GitHub doesn't allow me to add this comment outside of the PR diff).

reykjalin replied:
Ah yeah, I didn't notice that this was only for credit cards and SEPA payments 🤔 

I've applied this change in d56044fc47db5e3c6d99cde894d1610307571e66 🙌

ricardo commented on client/blocks/credit-card/use-payment-processing.js:
```diff
@@ -114,17 +111,24 @@ export const usePaymentProcessing = (
-							stripe_source: response.source.id,
+							stripe_source:
+								response?.source?.id ??
+								response?.paymentMethod?.id,
```
Looks good as is, so feel free to ignore 👇

We could simplify this a bit by assigning the source/paymentMethod id to a variable:

```JavaScript
const source_id = response?.source?.id ?? response?.paymentMethod?.id;

if ( ! source_id ) {
	throw new Error(
		getErrorMessageForTypeAndCode( errorTypes.API_ERROR )
	);
}

setSourceId( source_id );

// ...
```

reykjalin replied:
Nice, I like this approach more! Applied in 6f1dd391803ecfe25fddb3dbcce42a17faa9b713

ricardo replied:
@reykjalin it'd be nice to add it to `stripe_source:` below as well.

reykjalin replied:
Fixed in b85b10b1ed5e8a4f1908e3870b5a775d3dc45214. Nice catch, I completely missed that :)

ricardo commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -561,6 +568,10 @@ public function process_response( $response, $order ) {
 			$this->update_fees( $order, is_string( $response->balance_transaction ) ? $response->balance_transaction : $response->balance_transaction->id );
 		}
 
+		if ( isset( $response->payment_method_details ) && isset( $response->payment_method_details->card ) && isset( $response->payment_method_details->card->mandate ) ) {
```
Maybe we can simplify this a little:

```suggestion
		if ( isset( $response->payment_method_details->card->mandate ) ) {
```

reykjalin replied:
Nice, I didn't know this was possible! 🎉

reykjalin replied:
Applied in d13134b491a3a9e0eaaecd59e190f17879767346

ricardo commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1486,6 +1513,16 @@ public function save_intent_to_order( $order, $intent ) {
 		}
 
+		// Add the mandate id necessary for renewal payments with Indian cards.
+		if ( isset( $intent->charges ) && isset( $intent->charges->data ) ) {
```
Here we can also use just one `isset`:

```suggestion
		if ( isset( $intent->charges->data ) ) {
```

reykjalin replied:
Applied in dbad167a07b43a1ad3203b0fccea7322272407c7

ricardo commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1486,6 +1513,16 @@ public function save_intent_to_order( $order, $intent ) {
+		if ( isset( $intent->charges ) && isset( $intent->charges->data ) ) {
+			$charge = end( $intent->charges->data );
+
+			if ( isset( $charge->payment_method_options ) && isset( $charge->payment_method_options->card ) && isset( $charge->payment_method_options->card->mandate ) ) {
```
```suggestion
			if ( isset( $charge->payment_method_options->card->mandate ) ) {
```

reykjalin replied:
Applied in dbad167a07b43a1ad3203b0fccea7322272407c7

ricardo commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -664,7 +670,7 @@ public function save_payment_method( $source_object ) {
 		$customer = new WC_Stripe_Customer( $user_id );
 
-		if ( ( $user_id && 'reusable' === $source_object->usage ) ) {
+		if ( ( $user_id && ( substr( $source_object->id, 0, 3 ) === 'pm_' || 'reusable' === $source_object->usage ) ) ) {
```
```
substr( $source_object->id, 0, 3 ) === 'pm_'
```

This is repeating several times now. I'm wondering if we should export this to a `is_payment_method_token` `WC_Stripe_Helper` function or something? Perhaps too simple to even bother. Just commenting as a loud thought. No need to consider it 🙂

reykjalin replied:
I've been thinking the same thing actually 😅 

I'll make the change tonight when I go through the whole review 🙂 
Shouldn't take long and will be worth it in the long run 👍

reykjalin replied:
Added helper functions in 150dcee813e739bb122038a2f25cac5e3ca8f0cb

ricardo reviewed (changes requested):
Looking good! 👍 

Just left a few minor comments. Testing is pretty extensive, so I still didn't finish it, but I'll provide feedback when I do.

LevinMedia reviewed (dismissed):
Approved with the comments I left above.

ricardo commented:
@reykjalin An update on my review:

The code/approach is looking good to me.

I didn't get too deep into the testing. It's a lot of stuff to test, multiplied by different cards, multiplied by UPE vs non-UPE, classic vs blocks checkout and a few other nuances. - I didn't test any of the flows with UPE enabled.

I have two additional comments:

1. We might need to add the mandate id logic on pre-orders as well.
2. I found an issue when renewing a subscription with price different than parent order price:
  **Steps to reproduce:**
    1. As a shopper, add subscription with sign up fee or subscription + simple product to the cart.
    2. Purchase subscription using India 3DS card.
    3. Try renewing that subscription. Stripe will throw the following error
      > payment_intent_mandate_invalid
      > This mandate can only be used for a payment of X. You attempted a payment of Y.

Commit:
Update note informing merchant of off-session auth

Commit:
Use ! empty instead of isset && !== ''

Makes the if statements much easier to read.

Commit:
Simplify card handling flow in assets/js/stripe.js

Commit:
Refactor source creation in paymentProcessing hook

Commit:
Simplify check for mandate ID in API response

Commit:
Simplify check for mandate ID in payment intent

Commit:
Add helper function to evaluate payment methods

Helper methods allow us to determine:
 * Is a given ID a payment method?
 * Is a given object a payment method?
 * Is a given source/payment method reusable?

Commit:
Fix error when adding mandate info to API request

There was the possibility of an issue when $request['confirm'] is not
set.

Commit:
Add missing mandate information to UPE payments

Commit:
Fix comment on unused param for UPE mandate info

Commit:
Fix missing assignment for UPE API request

Commit:
Use correct amount when order has more than just a sub

reykjalin commented:
> I found an issue when renewing a subscription with price different than parent order price:

Nice catch! 🤩 

@ricardo I think I managed to fix the issue in 864cbb8a40dd4f76dc4c6feea40cad83f8e64f05. I wasn't able to confirm it since there seems to be some problem with the testing API — it's not creating mandates at the moment even though I'm passing it the right information. Not sure why.

> We might need to add the mandate id logic on pre-orders as well.

Hmmm, I don't think so since it's not a recurring payment. Since it's a pre-order I think you'd usually pay upfront too?

**EDIT:** I've confirmed this is an issue with the API. A commit that was working yesterday is no long working.

Commit:
Improve date format for order note

ismaeldcom commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -62,6 +65,88 @@ public function maybe_init_subscriptions() {
+		$order_amount = WC_Stripe_Helper::get_stripe_amount( $order->get_total() );
+		if ( $has_sign_up_fee || $sub_amount !== $order_amount ) {
+			$request['payment_method_options']['card']['mandate_options']['amount_type'] = 'maximum';
+		}
```
🤔 Knowing that using `$sub->get_total()` will return the recurring amount. Maybe I'm wrong, but I think there is no need to use the `maximum` type because the subscription has a sign-up fee or the order includes other products. Should be enough to just use the sum of subs amounts as the fixed amount on the mandate.

reykjalin replied:
Hmmm, I thought the first (i.e. on-session) payment had to fit the mandate requirements as well, but maybe not? I'll test this as soon as the testing APIs are working as intended.

Nice catch @ismaeldcom 🙌

reykjalin replied:
You're right, we can just use the `$sub_amount` here, even if there are other items in the cart :)

Nice catch! 👏

Commit:
update built JS

Commit:
Fix a revert accidentally done during a refactor

The order note went back to the previous version, now uses the proper
phrasing.

Commit:
Always use the sub total amount for mandate amount

This gets around a bunch of edge cases, e.g. when there are other items
in the cart.

diegocurbelo commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -678,7 +689,11 @@ public function get_source_object( $source_id = '' ) {
+		if ( WC_Stripe_Helper::is_id_for_payment_method( $source_id ) ) {
+			$source_object = WC_Stripe_API::retrieve( 'payment_methods/' . $source_id );
+		} else {
+			$source_object = WC_Stripe_API::retrieve( 'sources/' . $source_id );
```
This was replaced in other places with: `$this->get_source_object( $source )`... but I'm ok with keeping it to better match the `payment_methods retrieve` used above 🤷🏼‍♂️

reykjalin replied:
This is in the `get_source_object` function 🙂 

I decided to treat the Payment Method as a source just to minimize any changes to the code — in practice they're pretty interchangeable, so I thought it'd be best to keep the changes to a minimum and just retrieve the right object based on the ID here 🤔

diegocurbelo replied:
I'm dumb 🤦🏼🤦🏼‍♂️🤦🏼‍♀️... I really need the Picard facepalm for this

reykjalin replied:
Hahah not at all! It's easy to miss when the function name isn't part of the diff ☺️

diegocurbelo commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -541,7 +541,13 @@ public function process_payment( $order_id, $retry = true, $force_save_source =
+			) {
 				$upe_gateway = new WC_Stripe_UPE_Payment_Gateway();
 				return $upe_gateway->process_payment_with_saved_payment_method( $order_id );
 			}
```
Yeah, we can remove this... when UPE is enabled the `process_payment()` on the UPE gateway is used instead of this one, so if we are executing this method the `if` with `WC_Stripe_Feature_Flags::is_upe_checkout_enabled()` should always be always false anyway.

_As a side note, this was added as a workaround to allow saved payment methods added with UPE enabled to be used even if UPE was disabled by using the UPE Gateway anyway... prior to this "fix", we just removed them when disabling UPE._

reykjalin replied:
I'll try removing this, save a card with UPE, disable UPE, and try using the saved card with the classic checkout experience before I commit to removing this 👍

reykjalin replied:
Using a card saved with UPE worked 🎉 

I pushed the change in 2903f40b1b55d9eb2127a4182a5532cd58595188 👍

Commit:
Remove unnecessary check for cards sved with UPE

This was necessary when the non-UPE elements didn't support Payment
Methods, but now that PMs are supported it's safe to remove this check
and have the non-UPE process_payment function handle those cards as well

diegocurbelo reviewed (dismissed):
Looks good and tests well, nice job 🎉 ...
We can continue with the known issues in different PRs 🚢

cesarcosta99 commented:
I performed all tests listed above with the card numbers you provided and they all went fine :tada:

With non-Blocks checkout and UPE disabled I could not make a purchase. It can be something wrong on my end, so please ensure someone else can cover that.

I went through the critical flows with the non-Blocks checkout and UPE enabled (using `4242 4242 4242 4242`), here's what I got:

- I couldn't test **Checkout with Payment Request** as the buttons were not visible, even when I switched to Google Chrome. I also have enabled Payment Request Buttons. I never tested express checkouts so I don't know if there are extra steps to be taken, if someone already has it set up would be nice to cover it as well.
- I did not perform tests that require access to the Stripe account, since I'm using yours, such as manual capture, refunds etc.
- I found an issue on [Change payment method to new card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#change-payment-method-to-new-card), a fatal error after filling in the new payment method fields and submit:
```Fatal error: Uncaught Error: Call to undefined method WC_Stripe_UPE_Payment_Gateway::create_token_from_setup_intent() in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php on line 297```
- In the **Renew subscription manually** test, we are supposed to see two buttons in the subscription view but I saw three (Cancel, Change payment and Renew now). I guess **Change payment** should not display in this case? If it's a bug I can create an issue.

I could reproduce the fatal error consistently, let me know if you have trouble with it.

Commit:
Use source ID variable where applicable

I failed to add it to one field where it helps simplify the code.

reykjalin commented:
Thank you for testing @cesarcosta99 !

> I couldn't test Checkout with Payment Request as the buttons were not visible, even when I switched to Google Chrome. I also have enabled Payment Request Buttons. I never tested express checkouts so I don't know if there are extra steps to be taken, if someone already has it set up would be nice to cover it as well.

For the PRBs to load you must do the following:
- For Apple Pay: Add a card to the Apple Pay wallet on your iPhone/Mac. Open the store in Safari. **Requires real card; card won't be charged in test mode**
- For Google Pay: Use Google Chrome, log into your Google Pay account. Open the store in Google Chrome. **Requires real card; card won't be charged in test mode**
- For a generic payment request: Use a chromium based browser (Google Chrome, Brave, etc.), save card information in the AutoFill settings in your browser, use that browser to open the store. **You can add Stripe's test cards here**

***

> I found an issue on Change payment method to new card, a fatal error after filling in the new payment method fields and submit

Nice catch! This, however, is unrelated to this PR. We opened a separate issue for this in #2078 🙂 

***

> In the Renew subscription manually test, we are supposed to see two buttons in the subscription view but I saw three (Cancel, Change payment and Renew now). I guess Change payment should not display in this case? If it's a bug I can create an issue.

Hmmm, this seems like a problem with the testing instructions 🤔 
Yes, I think it'd be a good idea to open an issue for this. Please tag it with `priority: high` 🙂

reykjalin commented:
Superceded by https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2561
